### PR TITLE
chore: add protected trpc router example

### PR DESCRIPTION
**Pull request by bhatvikrant**
# Add protected trpc router example

- [x] I reviewed linter warnings + errors, resolved formatting, types and other issues related to my work
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

# Short summary
closes https://github.com/t3-oss/create-t3-app/issues/191
---

### CODE REVIEW

**Comment by bhatvikrant** [comment]
@nexxeln @TheoBr @juliusmarminge

File: template/addons/next-auth/protected-router.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { createRouter } from "./context";
+
+export function createProtectedRouter() {
+  return createRouter().middleware(({ ctx, next }) => {
+    if (!ctx.session || !ctx.session.user) {
+      throw new trpc.TRPCError({ code: "UNAUTHORIZED" });
+    }
+    return next({
+      ctx: {
+        ...ctx,
```
Comment (juliusmarminge) [inline template/addons/next-auth/protected-router.ts:11]: Didn't Alex say this wasn't needed or was that new in tRPC v10?

**Comment by juliusmarminge** [comment]
Also we have this already, probably best to do something about this as well to avoid confusion on where to put stuff:
https://github.com/t3-oss/create-t3-app/blob/da461a2b4e6d25d0989d17e89cacf7ce0aac8d2d/template/addons/trpc/auth-router.ts#L4-L22

File: template/addons/next-auth/protected-router.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import * as trpc from "@trpc/server";
+import { createRouter } from "./context";
+
+export function createProtectedRouter() {
```
[the remark this answers is not in the document]
Comment (bhatvikrant) [reply template/addons/next-auth/protected-router.ts:4]: what do you mean by `tRPC installer` ?

[the remark this answers is not in the document]
Comment (bhatvikrant) [reply template/addons/next-auth/protected-router.ts:4]: Got it

[the remark this answers is not in the document]
Comment (bhatvikrant) [reply template/addons/next-auth/protected-router.ts:4]: Should I re-create the `protected-router.ts` in `template/addons/trpc`? or is there a way to share the file?

File: src/installers/next-auth.ts [hunk 3]
```diff
@@ -41,9 +41,16 @@ export const nextAuthInstaller: Installer = async ({
...
+  const protectedRouterDest = path.join(
+    projectDir,
+    "src/server/router/protected-router.ts",
+  );
+
   await Promise.all([
     fs.copy(apiHandlerSrc, apiHandlerDest),
     fs.copy(restrictedApiSrc, restrictedApiDest),
     fs.copy(nextAuthDefinitionSrc, nextAuthDefinitionDest),
+    fs.copy(protectedRouterSrc, protectedRouterDest),
```
[the remark this answers is not in the document]
Comment (bhatvikrant) [reply src/installers/next-auth.ts:54]: Changing it to: 
```ts
...
    packages?.trpc.inUse ? fs.copy(protectedRouterSrc, protectedRouterDest): [],
...
```

File: template/addons/next-auth/protected-router.ts (hunk 1 above)
Comment (bhatvikrant) [reply template/addons/next-auth/protected-router.ts:11]: Not sure about v10, yet to see the updates.

But Theo asked for this [here](https://github.com/t3-oss/create-t3-app/issues/191)

@juliusmarminge

**Comment by bhatvikrant** [comment]
> Also we have this already, probably best to do something about this as well to avoid confusion on where to put stuff:
> 
> https://github.com/t3-oss/create-t3-app/blob/da461a2b4e6d25d0989d17e89cacf7ce0aac8d2d/template/addons/trpc/auth-router.ts#L4-L22

removing this middleware @juliusmarminge

File: template/addons/next-auth/protected-router.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply template/addons/next-auth/protected-router.ts:4]: Put it only in the trpc installer. Its a trpc router

File: template/addons/trpc/auth-index-router.ts [hunk 4]
```diff
@@ -3,12 +3,10 @@ import { createRouter } from "./context";
 import superjson from "superjson";
 
 import { exampleRouter } from "./example";
-import { authRouter } from "./auth";
 
 export const appRouter = createRouter()
   .transformer(superjson)
-  .merge("example.", exampleRouter)
-  .merge("auth.", authRouter);
```
Comment (juliusmarminge) [inline template/addons/trpc/auth-index-router.ts:11]: Wouldn't it be nice though to keep an example of a protected route using the protected router?

**Comment by bhatvikrant** [comment]
> Looks good. Left a final thought to think about before merging.

Noted @juliusmarminge 
I am travelling right now, will add an example on monday, please hold this PR until then.

File: template/addons/trpc/auth-index-router.ts (hunk 4 above)
Comment (TheoBr) [reply template/addons/trpc/auth-index-router.ts:11]: +1 to this, looks great otherwise!

**Comment by juliusmarminge** [comment]
> > Looks good. Left a final thought to think about before merging.
> 
> Noted @juliusmarminge 
> I am travelling right now, will add an example on monday, please hold this PR until then.

Have you had time to add an example or should i merge this and leave it for now?

**Comment by bhatvikrant** [comment]
> > > Looks good. Left a final thought to think about before merging.
> > 
> > 
> > Noted @juliusmarminge
> > I am travelling right now, will add an example on monday, please hold this PR until then.
> 
> Have you had time to add an example or should i merge this and leave it for now?

Will be adding the example, pls give a few hours more

**Comment by nexxeln** [comment]
> > > > Looks good. Left a final thought to think about before merging.
> > > 
> > > 
> > > Noted @juliusmarminge
> > > I am travelling right now, will add an example on monday, please hold this PR until then.
> > 
> > 
> > Have you had time to add an example or should i merge this and leave it for now?
> 
> Will be adding the example, pls give a few hours more

No rush 🙂

**Review by nexxeln — APPROVED** [review]
Looks good

File: src/installers/trpc.ts [hunk 5]
```diff
@@ -61,11 +55,30 @@ export const trpcInstaller: Installer = async ({
...
+    "src/server/router/protected-router.ts",
+  );
+
   await Promise.all([
     fs.copy(apiHandlerSrc, apiHandlerDest),
     fs.copy(utilsSrc, utilsDest),
     fs.copy(contextSrc, contextDest),
     fs.copy(indexRouterSrc, indexRouterDest),
     fs.copy(exampleRouterSrc, exampleRouterDest),
+    ...(usingAuth
```
Comment (juliusmarminge) [inline src/installers/trpc.ts:79]: Merge the two ternaries?

File: template/addons/trpc/protected-question-router.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { createProtectedRouter } from "./protected-router";
+
+export const protectedQuestionRouter = createProtectedRouter().query(
```
Comment (juliusmarminge) [inline template/addons/trpc/protected-question-router.ts:3]: Kind of feel like the previous "getSession" endpoint gave more value than just a mock-query?

Comment (bhatvikrant) [reply template/addons/trpc/protected-question-router.ts:3]: Fair enough

**Comment by bhatvikrant** [comment]
@juliusmarminge made the suggested changes

File: template/addons/trpc/protected-question-router.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { createProtectedRouter } from "./protected-router";
+
+export const protectedQuestionRouter = createProtectedRouter()
```
Comment (TheoBr) [inline template/addons/trpc/protected-question-router.ts:3]: nit: `Question` might confuse people as to "what this does", maybe `protectedExampleRouter`?

File: template/addons/trpc/protected-question-router.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { createProtectedRouter } from "./protected-router";
+
```
Comment (TheoBr) [inline template/addons/trpc/protected-question-router.ts:2]: Throw a comment on top explaining what this is an example "of", maybe?

File: template/addons/trpc/protected-router.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import * as trpc from "@trpc/server";
+import { createRouter } from "./context";
+
+export function createProtectedRouter() {
```
Comment (TheoBr) [inline template/addons/trpc/protected-router.ts:4]: Throw a comment on top of this describing what it is if you can - ideally tsdoc style so it has good info on hover 🙏

**Review by TheoBr — COMMENTED** [review]
Sorry for being pushy on code comments, I swear I don't give a shit normally but for a boilerplate-y thing like this I think it's important

**Comment by bhatvikrant** [comment]
> Sorry for being pushy on code comments, I swear I don't give a shit normally but for a boilerplate-y thing like this I think it's important

All are fair asks! 
Fixing them right away 🚀

**Comment by bhatvikrant** [comment]
I've made the suggested changes.
// @TheoBr

File: template/addons/trpc/protected-example-router.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { createProtectedRouter } from "./protected-router";
+
+// Create queries & mutations with the surety that user has a session (due to createProtectedRouter).
```
Comment (TheoBr) [inline template/addons/trpc/protected-example-router.ts:3]: // Example router with queries that can only be hit if the user requesting is signed in

File: template/addons/trpc/protected-router.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import * as trpc from "@trpc/server";
+import { createRouter } from "./context";
+
+/**
+ * Modify the default trpc router to use the next-auth context and check for user session.
```
Comment (TheoBr) [inline template/addons/trpc/protected-router.ts:5]: nit:
> Creates a tRPC router that asserts all queries and mutations are from an authorized user. Will throw an `unauthorized` error if a user is not signed in

**Comment by TheoBr** [comment]
Ty for getting to it so quick @bhatvikrant - left some thoughts on the comments themselves. Change those up and should be good to merge 🎉

**Comment by bhatvikrant** [comment]
Done @TheoBr

**Review by TheoBr — APPROVED** [review]

**Comment by juliusmarminge** [comment]
Thanks for your patience with all "fine-tune-requests".

**Pull request merged**